openedx/core/djangoapps/course_apps/plugins.py

```diff
@@ -0,0 +1,112 @@
+"""
+Course Apps plugin base class and plugin manager.
+"""
+from typing import Dict, Iterator, Optional
+
+from abc import ABC
+from edx_django_utils.plugins import PluginManager
+from opaque_keys.edx.keys import CourseKey
+
+
+# Stevedore extension point namespaces
+COURSE_APPS_PLUGIN_NAMESPACE = "openedx.course_app"
+
+
+class CourseApp(ABC):
+    """
+    Abstract base class for all course app plugins.
+    """
+
+    # A unique ID for the app.
+    app_id: str = ""
+    # A friendly name for the app.
+    name: str = ""
+    # A description for the app.
+    description: str = ""
+
+    @classmethod
+    def is_available(cls, course_key: CourseKey) -> bool:
+        """
+        Returns a boolean indicating whether or not this course app is available for a given course.
+
+        If an app is not available, it will not show up in the UI at all for that
+        course and it will not be possible to enable/disable/configure it.
+
+        Args:
+            course_key (CourseKey): Course key for course whose availability is being checked.
+
+        Returns:
+            bool: Availability status of app.
+        """
+        raise NotImplementedError()
+
+    @classmethod
+    def is_enabled(cls, course_key: CourseKey) -> bool:
+        """
+        Return if this course app is enabled for the provided course.
+
+        Args:
+            course_key (CourseKey): The course key for the course you
+                want to check the status of.
+
+        Returns:
+            bool: The status of the course app for the specified course.
+        """
+        raise NotImplementedError()
```

Comment:
we can update all these methods to "abstractmethods" as they are must-haves for the child class to implement.